### PR TITLE
Upload files

### PR DESCRIPTION
**Pull request by MarcelGarus**
Uploading files works now.

file screen with upload FAB | file picker dialog
---|---
file screen with upload FAB | file picker dialog

file picker menu | selecting multiple files | upload happening
--- | --- | ----
file picker menu | selecting multiple files | upload happening

**iOS is not setup yet**
An iOS developer would need to follow [these instructions](https://github.com/miguelpruivo/flutter_file_picker/wiki/Setup) (looking at @janrenz)

### CODE REVIEW

**Comment by janrenz** [comment]
Nice, if I upload from  the files home page, where do I upload?

**Comment by MarcelGarus** [comment]
Currently, the files always get put into the top folder of personal files, but the upload function already accepts a parent and owner, so it could theoretically be put anywhere.

File: lib/app/services/network.dart [hunk 1]
```diff
@@ -70,122 +68,108 @@ class TooManyRequestsError extends ServerError {
...
+/// servers.
 @immutable
 class NetworkService {
-  const NetworkService({@required this.apiUrl}) : assert(apiUrl != null);
+  const NetworkService();
+
+  /// Calls the given function and turns various status codes and socket
+  /// exceptions into custom error types like [AuthenticationError] or
+  /// [NoConnectionToServerError].
+  Future<http.Response> _makeCall(Future<http.Response> Function() call) async {
```
Comment (JonasWanke) [inline lib/app/services/network.dart]: When merging master into this branch, please make sure to re-add the logging statement and also pass a string of the HTTP method (https://github.com/schul-cloud/schulcloud-flutter/blob/19c0dd1b2cabcbe339f76dd6d04fb15059177e35/lib/app/services/network.dart#L141-L191)

File: lib/app/services/api_network.dart [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import 'package:flutter/widgets.dart';
+import 'package:get_it/get_it.dart';
+import 'package:meta/meta.dart';
+import 'package:http/http.dart' as http;
+import 'package:schulcloud/app/app.dart';
+
+import 'storage.dart';
+
+/// A service that offers networking POST and GET requests to the backend
```
Comment (JonasWanke) [inline lib/app/services/api_network.dart]: Only POST and GET requests is no longer correct—maybe just say "network requests"?

File: lib/app/services/api_network.dart [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  NetworkService get _network => services.get<NetworkService>();
+
+  Map<String, String> _getHeaders() {
+    final storage = services.get<StorageService>();
+    return {
+      'Content-Type': 'application/json',
+      if (storage.hasToken)
+        'Authorization': 'Bearer ${storage.token.getValue()}',
+    };
```
Comment (JonasWanke) [inline lib/app/services/api_network.dart]: `StorageService` is available as `services.storage`
```suggestion
    return {
      'Content-Type': 'application/json',
      if (storage.hasToken)
        'Authorization': 'Bearer ${services.storage.token.getValue()}',
    };
```

File: lib/app/services/network.dart [hunk 4]
```diff
@@ -70,122 +68,108 @@ class TooManyRequestsError extends ServerError {
   final Duration timeToWait;
 }
 
-/// A service that offers networking HTTP requests to the backend servers.
-///
-/// It depends on the authentication storage, so if the user's token
-/// is stored there, the requests' headers are automatically enriched with the
-/// access token.
+/// A service that offers networking POST and GET requests to arbitrary
```
Comment (JonasWanke) [inline lib/app/services/network.dart]: Only POST and GET requests is no longer correct—maybe just say "network requests"?

File: lib/app/services/api_network.dart [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import 'package:schulcloud/app/app.dart';
+
+import 'storage.dart';
+
+/// A service that offers networking POST and GET requests to the backend
+/// servers. If the user's token is stored in the authentication storage, the
+/// the requests' headers are automatically enriched with the access token.
+@immutable
+class ApiNetworkService {
+  const ApiNetworkService({@required this.apiUrl}) : assert(apiUrl != null);
```
Comment (JonasWanke) [inline lib/app/services/api_network.dart]: Why use a constructor parameter for the URL? You can retrieve it as `services.get<AppConfig>.baseApiUrl`

File: lib/file/widgets/upload_button.dart [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import 'package:flutter/material.dart';
+import 'package:schulcloud/app/app.dart';
+
+class UploadButton extends StatelessWidget {
+  const UploadButton({@required this.onPressed});
```
Comment (JonasWanke) [inline lib/file/widgets/upload_button.dart]: Missing an assert?

File: lib/file/bloc.dart [hunk 7]
```diff
@@ -97,4 +102,92 @@ class FileBloc {
...
+      snackbar.close();
+    }
+  }
+
+  Future<void> _uploadSingleFile({
+    @required io.File file,
+    @required Id<dynamic> owner,
+    Id<File> parent,
+  }) async {
+    if (file == null || !file.existsSync()) {
```
Comment (JonasWanke) [inline lib/file/bloc.dart]: Shouldn't `file != null` be an assertion?

File: lib/file/bloc.dart [hunk 8]
```diff
@@ -97,4 +102,92 @@ class FileBloc {
...
+                    : 'Uploading ${file.name} (${i + 1} / ${files.length})',
+              ),
+            ),
+          ],
+        ),
+      ));
+
+      await _uploadSingleFile(file: file, owner: owner, parent: parent);
+
+      snackbar.close();
```
Comment (JonasWanke) [inline lib/file/bloc.dart]: I really don't like this mix of widgets and logic. Suggestion: Return a stream with the necessary information to update the snack bar which is created somewhere in a widget file. (Note: consume the stream inside the snack bar so when the user dismisses it, they don't get new snack bars repeatedly, except the 'Done' message.) This is also missing L10n and a final snack bar when the upload is completed.

**Review by JonasWanke — CHANGES_REQUESTED** [review]

File: lib/file/bloc.dart (hunk 8 above)
Comment (JonasWanke) [reply lib/file/bloc.dart]: Maybe include a button in the final snack bar to open the folder uploaded to? And we should notify the user where they are uploading to beforehand.

Comment (MarcelGarus) [reply lib/file/bloc.dart]: I believe we manage to upload to the current folder in MVP MVP, so I won't add a localized string notifying the user about the upload folder.
The button should probably be put into another issue.

File: lib/app/services/api_network.dart [hunk 9]
```diff
@@ -36,7 +30,7 @@ class ApiNetworkService {
     Map<String, String> parameters = const {},
   }) {
     return _network.get(
-      _url(path),
+      scWebUrl(path),
```
Comment (JonasWanke) [inline lib/app/services/api_network.dart]: `scWebUrl` is for the web-client, not for the API. Please also change the three occurrences below

File: lib/file/bloc.dart [hunk 10]
```diff
@@ -114,70 +116,66 @@ class FileBloc {
     // Let the user pick files.
     final files = await FilePicker.getMultiFile();
 
+    final progressUpdates = BehaviorSubject<UploadProgressUpdate>();
```
Comment (JonasWanke) [inline lib/file/bloc.dart]: For better separation of widget and service layer, I suggest you transform this into an `async*`-method returning a `Stream` of progress updates. The UI can then take care of displaying the `SnackBar`.

File: lib/file/widgets/upload_snack_bars.dart [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+      ],
+    );
+  }
+}
+
+class UploadCompletedSnackBarContent extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) =>
+      Text(context.s.file_uploadCompletedSnackBar);
+}
```
Comment (JonasWanke) [inline lib/file/widgets/upload_snack_bars.dart]: Really‽

File: lib/sign_in/bloc.dart [hunk 12]
```diff
@@ -31,7 +31,9 @@ class SignInBloc {
       );
     }
 
-    // The login throws an exception if it wasn't successful.
+    logger.i('Logging in as $email…');
```
Comment (JonasWanke) [inline lib/sign_in/bloc.dart]:
```suggestion
    logger.i('Signing in as $email…');
```

**Review by JonasWanke — CHANGES_REQUESTED** [review]
Includes some strings that don't belong into this branch yet, but well…

**Review by JonasWanke — APPROVED** [review]

**Pull request merged**